add blog post on using kubectl annotate for human service discovery

richarddli opened the pull request:
We've found `kubectl annotate` to be a vastly underused and surprisingly useful part of Kubernetes, so I wrote a blog post about it. Feedback appreciated.

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+**Authors:** Richard Li (Ambassador Labs)
+
+Have you ever been asked to troubleshoot a failing Kubernetes service and struggled to find basic information about the service such as the source repository and owner?
```
Nice opening paragraph.

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+slug: annotating-k8s-for-humans
+---
+
+**Authors:** Richard Li (Ambassador Labs)
```
nit; I'd write:
```
**Author:** Richard Li,Ambassador Labs
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+One of the problems as Kubernetes applications grow is the proliferation of services. As the number of services grows, developers start to specialize working with specific services. When it comes to troubleshooting, however, developers need to be able to find the source, understand the service and dependencies, and chat with the owning team for any service.
+
+## Human Service Discovery
```
Our style guide prefers headings at HTML `<h2>` and higher to be in sentence case, so:
```suggestion
## Human service discovery
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+Troubleshooting always begins with information gathering. While much attention has been paid to centralizing machine data (e.g., logs, metrics), much less attention has been given to the human aspect of service discovery. Who owns a particular service? What Slack channel does the team work on? Where is the source for the service? What issues are currently known and being tracked?
+
+## Kubernetes Annotations
```
## Kubernetes annotations

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+## Kubernetes Annotations
+
+Kubernetes annotations are designed to solve exactly this problem. Oft-overlooked, Kubernetes annotations are designed to add metadata to Kubernetes objects. The Kubernetes documentation says annotations can “attach arbitrary non-identifying metadata to objects.” This means that annotations should be used for attaching metadata that is external to Kubernetes (i.e., metadata that Kubernetes won’t  ...
```
nit for URL format:
```suggestion
Kubernetes annotations are designed to solve exactly this problem. Oft-overlooked, Kubernetes annotations are designed to add metadata to Kubernetes objects. The Kubernetes documentation says annotations can “attach arbitrary non-identifying metadata to objects.” This means that annotations should be used for attaching metadata that is external to Kubernetes (i.e., metadata that Kubernetes won’t use to identify objects. As such, annotations can contain any type of data. This is a contrast to labels, which are designed for uses internal to Kubernetes. As such, label structure and values are [constrained](/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) so they can be efficiently used by Kubernetes.
```
**or** also link to annotations:
```suggestion
Kubernetes [annotations](/docs/concepts/overview/working-with-objects/annotations/) are designed to solve exactly this problem. Oft-overlooked, Kubernetes annotations are designed to add metadata to Kubernetes objects. The Kubernetes documentation says annotations can “attach arbitrary non-identifying metadata to objects.” This means that annotations should be used for attaching metadata that is external to Kubernetes (i.e., metadata that Kubernetes won’t use to identify objects. As such, annotations can contain any type of data. This is a contrast to labels, which are designed for uses internal to Kubernetes. As such, label structure and values are [constrained](/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) so they can be efficiently used by Kubernetes.
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+Kubernetes annotations are designed to solve exactly this problem. Oft-overlooked, Kubernetes annotations are designed to add metadata to Kubernetes objects. The Kubernetes documentation says annotations can “attach arbitrary non-identifying metadata to objects.” This means that annotations should be used for attaching metadata that is external to Kubernetes (i.e., metadata that Kubernetes won’t  ...
+
+
+## Kubernetes Annotations in Action
```
```suggestion
## Kubernetes annotations in action
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
````diff
@@ -0,0 +1,99 @@
+Here is an example. Imagine you have a Kubernetes service for quoting, called the quote service. You can do the following:
+
+```
+$ kubectl annotate service quote a8r.io/owner=”@sally”
````
_If you want_ you can follow the Kubernetes documentation style guide and omit the prompt:
```suggestion
kubectl annotate service quote a8r.io/owner=”@sally”
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
````diff
@@ -0,0 +1,99 @@
+
+In this example, we’ve just added an annotation called `a8r.io/owner` with the value of @sally. Now, we can use `kubectl describe` to get the information.
+
+```
````
*If you want* you can follow the Kubernetes documentation style guide and omit the prompt, then start a new block for the example output. For blog articles we don't enforce that part of the style guide.

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
````diff
@@ -0,0 +1,99 @@
+
+If you’re practicing GitOps (and you should be!) you’ll want to code these values directly into your Kubernetes manifest, e.g.,
+
+```
````
If you want, append `yaml` to the block opener (to make <tt>\`\`\`yaml</tt>) and the website will syntax-highlight the example manifest.

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
````diff
@@ -0,0 +1,99 @@
+    app: quote
+```
+
+## A Convention for Annotations
````
```suggestion
## A convention for annotations
```
**or**
```suggestion
## Ambassador convention for annotations
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+Adopting a common convention for annotations ensures consistency and understandability. Typically, you’ll want to attach the annotation to the service object, as services are the high-level resource that maps most clearly to a team’s responsibility. Namespacing your annotations is also very important. Here is one set of conventions, documented at [a8r.io](https://a8r.io), and reproduced below:
+
+| Annotation                                 | Description                                 |
```
If you want, you can add a [table caption](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#table-captions) for accessibility.

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+Adopting a common convention for annotations ensures consistency and understandability. Typically, you’ll want to attach the annotation to the service object, as services are the high-level resource that maps most clearly to a team’s responsibility. Namespacing your annotations is also very important. Here is one set of conventions, documented at [a8r.io](https://a8r.io), and reproduced below:
+
+| Annotation                                 | Description                                 |
+| ------------------------------------------ | ------------------------------------------- |
```
nit: either all the descriptions should end with a period, or none should.

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+| `a8r.io/dependencies`                      | Unstructured text describing the service dependencies for humans. |
+
+
+## Visualizing Annotations: Service Catalogs
```
```suggestion
## Visualizing annotations: service catalogs
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+## Visualizing Annotations: Service Catalogs
+
+As the number of microservices and annotations proliferate, using `kubectl describe` can get tedious. Moreover, kubectl describe requires every developer to have some direct access to the Kubernetes cluster. A [service catalog](https://www.getambassador.io/learn/kubernetes-glossary/service-catalog) presents an internal, developer-oriented view of all services. 
```
I would write:
```suggestion
As the number of microservices and annotations proliferate, running `kubectl describe` can get tedious. Moreover, using `kubectl describe` requires every developer to have some direct access to the Kubernetes cluster. An [service catalog](https://www.getambassador.io/learn/kubernetes-glossary/service-catalog), in the Ambassador meaning of that term, presents an internal, developer-oriented view of all services.
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+As the number of microservices and annotations proliferate, using `kubectl describe` can get tedious. Moreover, kubectl describe requires every developer to have some direct access to the Kubernetes cluster. A [service catalog](https://www.getambassador.io/learn/kubernetes-glossary/service-catalog) presents an internal, developer-oriented view of all services. 
+
+Historically, service catalogs have not been a key piece of infrastructure, as applications consisted of a single monolith. Today, with distributed development and microservices the norm, service catalogs are increasingly being adopted as a best practice. Some popular options today include Backstage and Ambassador. [Backstage](https://backstage.io/) is a highly flexible project ideal for organiza ...
```
You could write:
Historically, service catalogs have not been a key piece of infrastructure, as applications consisted of a single monolith. Today, with distributed development and microservices the norm, service catalogs are increasingly being adopted as a best practice. Some popular options today include [Backstage](https://backstage.io/) and [Ambassador Service Catalog](https://www.getambassador.io/products/service-catalog). Backstage is a highly flexible project ideal for organizations that wish to customize every aspect of their service catalog, while Ambassador Service Catalog provides a turnkey, cloud-hosted solution that can be set up quickly.
```
I'd like to see some text to explain that this concept of a service catalog is different from Kubernetes [Service Catalog](https://svc-cat.io/).

richarddli replied:
Thanks! Your feedback is great. I pushed some changes; please let me know your thoughts.

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+Historically, service catalogs have not been a key piece of infrastructure, as applications consisted of a single monolith. Today, with distributed development and microservices the norm, service catalogs are increasingly being adopted as a best practice. Some popular options today include Backstage and Ambassador. [Backstage](https://backstage.io/) is a highly flexible project ideal for organiza ...
+
+## Annotate Your Services Now and Thank Yourself Later
```
This could stay in title case to make a point, or you could write:
```suggestion
## Annotate your services now and thank yourself later
```

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+Historically, service catalogs have not been a key piece of infrastructure, as applications consisted of a single monolith. Today, with distributed development and microservices the norm, service catalogs are increasingly being adopted as a best practice. Some popular options today include Backstage and Ambassador. [Backstage](https://backstage.io/) is a highly flexible project ideal for organiza ...
+
+## Annotate Your Services Now and Thank Yourself Later
+Much like implementing observability within microservice systems, you often don’t realize that you need human service discovery until it’s too late. Typically, something is on fire in production and you wished you had implemented better metrics and also specified which Slack channel the owning team lived in within your organization. 
```
How about:
```suggestion
Much like implementing observability within microservice systems, you often don’t realize that you need human service discovery until it’s too late. Don't wait until something is on fire in production to start wishing you had implemented better metrics and also documented how to get in touch with the part of your organization that looks after it.
```
?

sftim reviewed (dismissed):
Hi @richarddli 

I really like the idea of an article on this topic. I have one strong concern and all the other feedback is minor things. The strong concern is to make sure that readers who've heard a bit about https://github.com/kubernetes-sigs/service-catalog/ and https://svc-cat.io/ don't get mixed up and think those are the same as Service Catalog from Ambassador.

Would you be willing to revise this PR to make that clear?

richarddli commented:
@sftim Thank you for the excellent feedback. I incorporated all of your suggestions and added some more explanation of the difference between the Kubernetes service catalog and the general "Category" of service catalogs. It's a little confusing, but the term service catalog has at this point been fairly well popularized by the Spotify team in particular, so I'm glad you called this out.

richarddli commented:
@sftim @kbarnard10 Hello! Just a friendly poke to see if you had any other feedback. Thanks! 😀

kbarnard10 commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+
+As the number of microservices and annotations proliferate, running `kubectl describe` can get tedious. Moreover, using `kubectl describe` requires every developer to have some direct access to the Kubernetes cluster. Over the past few years, [service catalogs](https://www.getambassador.io/learn/kubernetes-glossary/service-catalog) have gained greater visibility in the Kubernetes ecosystem. Popul ...
+
+Historically, service catalogs have not been a key piece of infrastructure, as applications consisted of a single monolith. Today, with distributed development and microservices the norm, service catalogs are increasingly being adopted as a best practice. Some popular options today include [Backstage]((https://backstage.io/) and [Ambassador Service Catalog](https://www.getambassador.io/products/s ...
```
```suggestion
Historically, service catalogs have not been a key piece of infrastructure, as applications consisted of a single monolith. Today, with distributed development and microservices the norm, service catalogs are increasingly being adopted as a best practice. Some popular options today include [Backstage](https://backstage.io/) and [Ambassador Service Catalog](https://www.getambassador.io/products/service-catalog). Backstage is a highly flexible project ideal for organizations that wish to customize every aspect of their service catalog, while Ambassador Service Catalog provides a turnkey, cloud-hosted solution that can be set up quickly.
```

richarddli commented:
@bobbytables @kbarnard10 Just another poke; is there anything I need to do here? Thanks! 😀

richarddli commented:
@bobbytables @kbarnard10 hello, just checking in on this!

sftim reviewed (commented):
Based on a quick scan, the latest revision LGTM

I'd squash the commits @richarddli (but you _might_ want to wait until we've assigned a publication date).

celestehorgan commented:
@kbarnard10 – does this look ready to go from your end?

richarddli commented:
Thanks, @sftim! @kbarnard10 can you advise on anything else I need to do? Thanks!

richarddli commented:
@kbarnard10 ping again?

zacharysarah commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+---
+layout: blog
+title: 'Annotating Kubernetes Services for Humans'
+date: 2021-02-27
```
```suggestion
date: 2021-04-06
```

zacharysarah commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+## Annotate your services now and thank yourself later
+Much like implementing observability within microservice systems, you often don’t realize that you need human service discovery until it’s too late. Don't wait until something is on fire in production to start wishing you had implemented better metrics and also documented how to get in touch with the part of your organization that looks after it.
+
+There's enormous benefits to building an effective “[version 0](https://www.getambassador.io/learn/kubernetes-glossary/version-0/)” of any service: a “[dancing skeleton](https://containerjournal.com/topics/container-management/dancing-skeleton-apis-and-microservices/)” application with a thin slice of complete functionality that can be deployed to production with a minimal yet effective continuou ...
```
```suggestion
There's enormous benefits to building an effective [version 0](https://www.getambassador.io/learn/kubernetes-glossary/version-0/) of any service: a “[dancing skeleton](https://containerjournal.com/topics/container-management/dancing-skeleton-apis-and-microservices/)” application with a thin slice of complete functionality that can be deployed to production with a minimal yet effective continuous delivery pipeline. 
```

zacharysarah reviewed (changes requested):
@richarddli 👋🏻  Thanks for your patience! This looks good to me. ~and I'll merge it today after some slight tweaks for date and grammar.~

Update: the 1.21 release is this week, so this post needs to wait until after the release, let's say Monday the 12th. I'll change the date again.

Otherwise, LGTM/approve.

zacharysarah commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+---
+layout: blog
+title: 'Annotating Kubernetes Services for Humans'
+date: 2021-04-06
```
```suggestion
date: 2021-04-12
```

zacharysarah commented:
/lgtm
/approve

Explicit hold for blog owners: feel free to `/hold cancel` at any time, including before 12 April. @kbarnard10 @onlydole @mrbobbytables 
/hold

sftim commented:
@divya-mohan0209 is this publication date OK? It's not related to the v1.21 release

divya-mohan0209 commented:
> @divya-mohan0209 is this publication date OK? It's not related to the v1.21 release

@sftim Thank you so much for checking in. One of the Blog posts related to 1.21 is coming out on April 12th & there'll be a blogpost every alternate day after that till April 23rd. If we could shift this one to either the 13th/15th next week, that would be great.

richarddli commented:
Just checking on the scheduling of this -- fine with either 😀

sftim commented:
I think actually there might be a different article penciled in  for the 15th. Maybe the first free slot in the following week?

richarddli commented:
@sftim Sounds good! Can we get it locked please?

sftim commented on content/en/blog/_posts/2021-02-27-annotating-k8s-for-humans.md:
```diff
@@ -0,0 +1,99 @@
+---
+layout: blog
+title: 'Annotating Kubernetes Services for Humans'
+date: 2021-04-12
```
```suggestion
date: 2021-04-20
```

sftim commented:
With updated publication date from https://github.com/kubernetes/website/pull/26749#pullrequestreview-635009303

Echoing LGTM and approval from https://github.com/kubernetes/website/pull/26749#issuecomment-814359437
/lgtm
/approve

sftim commented:
/hold cancel

sftim commented:
@richarddli thanks for this article!

sftim commented:
Scheduled for automatic publication on 2021-04-20
